src/main/groovy/org/gradle/android/workarounds/RoomSchemaLocationWorkaround.groovy

```diff
@@ -148,27 +158,24 @@ class RoomSchemaLocationWorkaround implements Workaround {
                 }
 
                 task.finalizedBy onlyIfAnnotationProcessorConfiguredForKapt(annotationProcessorOptionProviders) { roomExtension.schemaLocationDir.isPresent() ? mergeTask : null }
+
+                TaskExecutionGraph taskGraph = task.project.gradle.taskGraph
+                taskGraph.whenReady onlyIfAnnotationProcessorConfiguredForKapt(annotationProcessorOptionProviders) { KaptRoomSchemaLocationArgumentProvider provider ->
+                    if (taskGraph.hasTask(task)) {
+                        // Register the variant-specific directory with the merge task
+                        roomExtension.registerOutputDirectory(provider.schemaLocationDir)
+                    }
+                }
             }
 
             project.tasks.withType(kaptWithoutKotlincTaskClass).configureEach(configureKaptTask)
             project.tasks.withType(kaptWithKotlincTaskClass).configureEach(configureKaptTask)
 
-            // Since we've added a new kapt-specific provider to the variant, go through the
-            // JavaCompile tasks and remove this provider from its task-specific list so that
-            // it only has its JavaCompile-specific provider.  This is not great, but there
+            // Since we've added a new kapt-specific provider to the variant, disable the provider
```

Comment:
I honestly am not sure why the javacompile task runs here. I think it might be a mistake if the kapt task is the one generating the schema. Do you think we should report this to Jetbrains?

Reply:
Hmm, I thought they were both compiled by the kotlin compiler since if a java file references a kotlin file, the java compiler would not know what to do with it.
For annotation processing this is achieved by the kotlin compiler creating java stubs for kotlin files but this is not done if there are no annotation processors.